File: 90.subsets-ii.cpp
/*
 * @lc app=leetcode id=90 lang=cpp
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> path;
        vector<vector<int>> ans;

        dfs(nums, 0, path, ans);
        return ans;
    }

    void dfs(vector<int>& nums, int pos, vector<int>& path, vector<vector<int>>& ans) {
        ans.push_back(path);
        if(pos == nums.size()) return;

        for(int i=pos; i<nums.size(); ++i) {
            if(i>pos and nums[i] == nums[i-1]) continue;
            path.push_back(nums[i]);
            dfs(nums, i+1, path, ans);
            path.pop_back();
        }
    }
};
// @lc code=end

File: 90.subsets-ii.py
#
# @lc app=leetcode id=90 lang=python3
#
# [90] Subsets II
#

# @lc code=start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        path = []
        ans = []

        def dfs(nums, pos):
            ans.append(path[:])
            if pos == len(nums):
                return
            
            for i in range(pos, len(nums)):
                if i>pos and nums[i] == nums[i-1]:
                    continue
                path.append(nums[i])
                dfs(nums, i+1)
                path.pop()

        dfs(nums, 0)
        return ans

    def subsetsSizeKWithDup(self, nums: List[int], k: int) -> List[List[int]]:
        nums.sort()
        path = []
        ans = []

        def dfs(nums, pos):
            if len(path) == k:
                ans.append(path[:])
                return
            if pos == len(nums):
                return
            
            for i in range(pos, len(nums)):
                if i>pos and nums[i] == nums[i-1]:
                    continue
                path.append(nums[i])
                dfs(nums, i+1)
                path.pop()

        dfs(nums, 0)
        return ans

    def subsetsNoNeighborWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        path = []
        ans = []

        def dfs(nums, pos):
            ans.append(path[:])
            if pos >= len(nums):
                return
            
            for i in range(pos, len(nums)):
                if i>pos and nums[i] == nums[i-1]:
                    continue
                path.append(nums[i])
                dfs(nums, i+2)
                path.pop()

        dfs(nums, 0)
        return ans

# @lc code=end

